File: app/src/test/java/com/papi/nova/KotlinDebugUiViewsMigrationTest.java
package com.papi.nova;

import android.content.Context;
import android.view.KeyEvent;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

import com.papi.nova.ui.AdapterFragment;
import com.papi.nova.ui.ExternalControllerView;

import org.junit.Test;

import java.io.File;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class KotlinDebugUiViewsMigrationTest {
    @Test
    public void debugAndUiViewsAreKotlinSources() {
        String[] names = {
                "DebugInfoActivity",
                "ui/AdapterFragment",
                "ui/ExternalControllerView"
        };

        for (String name : names) {
            File javaFile = new File("src/main/java/com/papi/nova/" + name + ".java");
            File kotlinFile = new File("src/main/java/com/papi/nova/" + name + ".kt");
            assertFalse(name + " should no longer be a Java source", javaFile.exists());
            assertTrue(name + " should be migrated to Kotlin", kotlinFile.exists());
        }
    }

    @Test
    public void migratedDebugAndUiViewsKeepJavaCompatibleApis() throws NoSuchMethodException {
        assertTrue(AppCompatActivity.class.isAssignableFrom(DebugInfoActivity.class));
        DebugInfoActivity.class.getConstructor();

        assertTrue(Fragment.class.isAssignableFrom(AdapterFragment.class));
        AdapterFragment.class.getConstructor();

        assertTrue(View.class.isAssignableFrom(ExternalControllerView.class));
        ExternalControllerView.class.getConstructor(Context.class);
        ExternalControllerView.class.getMethod("setInputCallbacks", ExternalControllerView.InputCallbacks.class);
        ExternalControllerView.class.getMethod("setCommitTextEnabled", boolean.class);

        assertTrue(ExternalControllerView.InputCallbacks.class.isInterface());
        ExternalControllerView.InputCallbacks.class.getMethod("handleKeyUp", KeyEvent.class);
        ExternalControllerView.InputCallbacks.class.getMethod("handleKeyDown", KeyEvent.class);
        ExternalControllerView.InputCallbacks.class.getMethod("handleCommitText", CharSequence.class);
        ExternalControllerView.InputCallbacks.class.getMethod("handleDeleteSurroundingText", int.class, int.class);
    }
}

File: app/src/test/java/com/papi/nova/KotlinDebugUiViewsMigrationTest.kt
package com.papi.nova

import android.content.Context
import android.view.KeyEvent
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import com.papi.nova.ui.AdapterFragment
import com.papi.nova.ui.ExternalControllerView
import java.io.File
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class KotlinDebugUiViewsMigrationTest {
    @Test
    fun debugAndUiViewsAreKotlinSources() {
        val names = arrayOf(
            "DebugInfoActivity",
            "ui/AdapterFragment",
            "ui/ExternalControllerView"
        )

        for (name in names) {
            val javaFile = File("src/main/java/com/papi/nova/$name.java")
            val kotlinFile = File("src/main/java/com/papi/nova/$name.kt")
            assertFalse("$name should no longer be a Java source", javaFile.exists())
            assertTrue("$name should be migrated to Kotlin", kotlinFile.exists())
        }
    }

    @Test
    fun migratedDebugAndUiViewsKeepJavaCompatibleApis() {
        assertTrue(AppCompatActivity::class.java.isAssignableFrom(DebugInfoActivity::class.java))
        DebugInfoActivity::class.java.getConstructor()

        assertTrue(Fragment::class.java.isAssignableFrom(AdapterFragment::class.java))
        AdapterFragment::class.java.getConstructor()

        assertTrue(View::class.java.isAssignableFrom(ExternalControllerView::class.java))
        ExternalControllerView::class.java.getConstructor(Context::class.java)
        ExternalControllerView::class.java.getMethod(
            "setInputCallbacks",
            ExternalControllerView.InputCallbacks::class.java
        )
        ExternalControllerView::class.java.getMethod(
            "setCommitTextEnabled",
            Boolean::class.javaPrimitiveType!!
        )

        assertTrue(ExternalControllerView.InputCallbacks::class.java.isInterface)
        ExternalControllerView.InputCallbacks::class.java.getMethod("handleKeyUp", KeyEvent::class.java)
        ExternalControllerView.InputCallbacks::class.java.getMethod("handleKeyDown", KeyEvent::class.java)
        ExternalControllerView.InputCallbacks::class.java.getMethod("handleCommitText", CharSequence::class.java)
        ExternalControllerView.InputCallbacks::class.java.getMethod(
            "handleDeleteSurroundingText",
            Int::class.javaPrimitiveType!!,
            Int::class.javaPrimitiveType!!
        )
    }
}
